proposals/NNNN-generalize-async-sequence.md

```diff
@@ -0,0 +1,311 @@
+
+#### Generalized `isolated` parameters
+
+Functions can abstract over actor isolation using `isolated` parameters. To enable representing a `nonisolated` context, `isolated` parameters can be made optional, where a `nil` argument value will cause the method to be dynamically `nonisolated`.
+
+Calling a function with an isolated parameter shares the isolation of the caller if:
+
+* the current context is non-isolated, the parameter type is optional, and the argument expression is `nil` or a reference to `Optional.none`;
+* the current context has an `isolated` parameter (including the implicitly-isolated `self` parameter of an actor function) and the argument expression is a reference to that parameter or a non-optional derivation of it (see below); or
+* the current context is isolated to a global actor type `T` and the argument expression is `T.shared`, where `shared` is `GlobalActor`'s protocol requirement or the concrete declaration which provides it in `T`'s conformance to `GlobalActor`.
+
+An expression is a non-optional derivation of an isolated parameter `param` if it is:
+
+* `param?` (the optional-chaining operator);
+* `param!` (the force-unwrapping operator); or
+* a reference to a _non-optional binding_ of `param`, i.e. a `let` constant initialized by a successful pattern-match which removes the optionality from `param`, such as `ref` in `if let ref = param`.
+
+When analyzing an argument expression in all cases above, certain non-instrumental differences in expression syntax and behavior must be ignored:
+
+* parentheses;
+* the effect-marking operators `try`, `try?`, `try!`, and `await`;
+* the type coercion operator as (in the cases where it doesn't perform a dynamic bridging conversion); and
+implicit type conversions such as promotion to `Optional` type.
+
+#### Calling `nextElement()` in async `for-in` loops
+
+Desugared async `for-in` loops will call `AsyncIteratorProtocol.nextElement()` instead of `next()` when the context has appropriate availability, and pass in an isolated argument value of type `(any Actor)?`. The argument value always represents the isolation of the caller so that the call does not cross an isolation boundary:
+
+* if the caller's context is `nonisolated`, the argument value is `nil`;
+* if the caller's context is isolated to some global actor `T`, the argument value is `T.shared`;
+* if the caller's context has an isolated parameter `param`, the argument value is `param`;
```

Comment:
I suppose this technically covers actor methods but might be good to spell it out explicitly

Reply:
I think we can now remove this whole list and say `#isolation` is used as the argument value. This list effectively describes what `#isolation` expands to, which is covered in the actor isolation inheritance proposal.